apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/link-follower-agent.ts

```diff
@@ -0,0 +1,406 @@
+      await filterAndRankTextChunksAgent({
+        description: resource.descriptionOfExpectedContent,
+        exampleText: resource.exampleOfExpectedContent,
+        textChunks,
+      });
+
+    if (
+      filteredAndRankedTextChunksResponse.status !== "ok" ||
+      filteredAndRankedTextChunksResponse.orderedRelevantTextChunks.length === 0
+    ) {
+      /** @todo: consider improving the error reporting of this */
+      return {
+        status: "not-explored",
+        reason: "No relevant sections found in the PDF.",
+        resource,
+      };
+    }
+
+    const { orderedRelevantTextChunks } = filteredAndRankedTextChunksResponse;
+
+    logger.debug(
+      `Ordered relevant text chunks: ${stringify(orderedRelevantTextChunks)}`,
+    );
+
+    content = dedent(`
+      Here is a list of the most relevant sections of the PDF file with file URL ${resource.url}:
+      ${orderedRelevantTextChunks
+        .map((text, index) => `Relevant section ${index + 1}: ${text}`)
+        .join("\n")}
+      `);
+  } else {
+    const webPage = await getWebPageActivity({
+      url: resource.url,
+      sanitizeForLlm: true,
+    });
+
+    logProgress([
+      {
+        recordedAt: new Date().toISOString(),
+        stepId,
+        type: "VisitedWebPage",
+        webPage: { url: webPage.url, title: webPage.title },
+        explanation: resource.reason,
+      },
+    ]);
+
+    content = webPage.htmlContent;
+  }
+
+  const { task, entityTypes, linkEntityTypes } = input;
+
+  const relevantLinksFromContent = await extractLinksFromContent({
+    contentUrl: resource.url,
+    content,
+    prompt: task,
+  }).then((response) => {
+    if (response.status === "ok") {
+      return response.links;
+    }
+
+    return [];
+  });
+
+  logger.debug(
+    `Extracted relevant links from the content of the resource with URL ${resource.url}: ${stringify(relevantLinksFromContent)}`,
+  );
+
+  const dereferencedEntityTypesById = {
+    ...entityTypes.reduce<DereferencedEntityTypesByTypeId>(
+      (prev, schema) => ({
+        ...prev,
+        [schema.$id]: { schema, isLink: false },
+      }),
+      {},
+    ),
+    ...(linkEntityTypes ?? []).reduce<DereferencedEntityTypesByTypeId>(
+      (prev, schema) => ({
+        ...prev,
+        [schema.$id]: { schema, isLink: true },
+      }),
+      {},
+    ),
+  };
+
+  const {
+    facts: inferredFactsFromContent,
+    entitySummaries: inferredFactsAboutEntitiesFromContent,
+  } = await inferFactsFromText({
+    text: content,
+    /** @todo: consider whether this should be a dedicated input */
+    relevantEntitiesPrompt: task,
+    dereferencedEntityTypes: dereferencedEntityTypesById,
+  });
+
+  const factSource: SourceProvenance = {
+    type: isResourcePdfFile ? SourceType.Document : SourceType.Webpage,
+    location: {
+      uri: resource.url,
+      /** @todo */
+      name: undefined,
+      description: undefined,
+    },
+    loadedAt: new Date().toISOString(),
+    /** @todo */
```

Comment:
Can you update these comments to specify how we might populate each?

Reply:
I've provided some ideas in improved todos as part of https://github.com/hashintel/hash/pull/4649/commits/72ee0992b97c863e52ac882f9efa3a0e2f6926b5, but in general this information will likely need to be extracted via an LLM if programmatic parsing method isn't available (or robust enough).